cmd/provisioner-localpv/app/helper_blockdevice.go

```diff
@@ -0,0 +1,214 @@
+	corev1 "k8s.io/api/core/v1"
+	//ndmv1alpha1 "github.com/openebs/maya/pkg/apis/openebs.io/ndm/v1alpha1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	bdcStorageClassAnnotation = "local.openebs.io/blockdeviceclaim"
+)
+
+//TODO
+var (
+	//WaitForBDTimeoutCounts specifies the duration to wait for BDC to be associated with a BD
+	//The duration is the value specified here multiplied by 5
+	WaitForBDTimeoutCounts = 12
+)
+
+// HelperBlockDeviceOptions contains the options that
+// will launch a BDC on a specific node (nodeName)
+type HelperBlockDeviceOptions struct {
+	nodeName string
+	name     string
+	capacity string
+	//	deviceType string
+	bdcName string
+}
+
+// validate checks that the required fields to create BDC
+// are available
+func (blkDevOpts *HelperBlockDeviceOptions) validate() error {
+	glog.Infof("Validate Block Device Options")
+	if blkDevOpts.name == "" || blkDevOpts.nodeName == "" {
+		return errors.Errorf("invalid empty name or node")
+	}
+	return nil
+}
+
+// hasBDC checks if the bdcName has already been determined
+func (blkDevOpts *HelperBlockDeviceOptions) hasBDC() bool {
+	glog.Infof("Already has BDC %t", blkDevOpts.bdcName != "")
+	return blkDevOpts.bdcName != ""
+}
+
+// getBlcokDeviceClaimFromPV inspects the PV and fetches the BDC associated
+//  with the Local PV.
+func (blkDevOpts *HelperBlockDeviceOptions) setBlockDeviceClaimFromPV(pv *corev1.PersistentVolume) error {
+	glog.Infof("Setting Block Device Claim From PV")
+	bdc, found := pv.Annotations[bdcStorageClassAnnotation]
+	if found {
+		blkDevOpts.bdcName = bdc
+	}
+	return nil
+}
+
+// createBlockDeviceClaim creates a new BlockDeviceClaim for a given
+//  Local PV
+func (p *Provisioner) createBlockDeviceClaim(blkDevOpts *HelperBlockDeviceOptions) error {
+	glog.Infof("Creating Block Device Claim")
+	if err := blkDevOpts.validate(); err != nil {
+		return err
+	}
+
+	if blkDevOpts.hasBDC() {
+		//already created
+		glog.Infof("Volume %v has been initialized with BDC:%v", blkDevOpts.name, blkDevOpts.bdcName)
+		return nil
+	}
+
+	//Setup the BDC Name using the provided PV name.
+	//To help easily co-relate, the name format for bdc will
+	//be  "bdc-<pvname>"
+	bdcName := "bdc-" + blkDevOpts.name
+
+	//TODO: Create BDC
+	bdcObj, err := blockdeviceclaim.NewBuilder().
+		WithNamespace(p.namespace).
+		WithName(bdcName).
+		WithHostName(blkDevOpts.nodeName).
+		WithCapacity(blkDevOpts.capacity).
+		Build()
+
+	if err != nil {
+		//TODO : Need to relook at this error
+		return errors.Errorf("unable to build BDC for pvc %v", blkDevOpts.name)
+	}
+
+	glog.Infof("Creating Block Device Claim using KubeClient")
+	_, err = blockdeviceclaim.NewKubeClient().
+		WithNamespace(p.namespace).
+		Create(bdcObj.Object)
+
+	if err != nil {
+		//TODO : Need to relook at this error
+		//If the error is about BDC being already present, then return nil
+		return errors.Errorf("unable to save BDC for pvc %v", blkDevOpts.name)
```

Comment:
let us wrap this error.
let the error message deal with what this overall function is supposed to do i.e. creating bdc
```go
errors.Wrapf(err, "failed to create BDC {%s}", blkDevOpts.name)
```